rustworkx-core/src/centrality.rs

```diff
@@ -672,6 +672,130 @@ where
+/// The implementation of this algorithm is based on the NetworkX
+/// [`katz_centrality()`](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.centrality.katz_centrality.html)
+/// function.
+///
+/// In the case of multigraphs the weights of any parallel edges will be
+/// summed when computing the eigenvector centrality.
+///
+/// Arguments:
+///
+/// * `graph` - The graph object to run the algorithm on
+/// * `weight_fn` - An input callable that will be passed the `EdgeRef` for
+///     an edge in the graph and is expected to return a `Result<f64>` of
+///     the weight of that edge.
+/// * `alpha` - Attenuation factor. If set to `None`, a default value of 0.1 is used.
+/// * `beta_map` - Immediate neighbourhood weights. Must contain all node indices or be `None`.
+/// * `beta_scalar` - Immediate neighbourhood scalar that replaces `beta_map` in case `beta_map` is None.
+///     Defaults to 1.0 in case `None` is provided.
+/// * `max_iter` - The maximum number of iterations in the power method. If
+///     set to `None` a default value of 100 is used.
+/// * `tol` - The error tolerance used when checking for convergence in the
+///     power method. If set to `None` a default value of 1e-6 is used.
+///
+/// # Example
+/// ```rust
+/// use rustworkx_core::Result;
+/// use rustworkx_core::petgraph;
+/// use rustworkx_core::petgraph::visit::{IntoEdges, IntoNodeIdentifiers};
+/// use rustworkx_core::centrality::katz_centrality;
+///
+/// let g = petgraph::graph::UnGraph::<i32, ()>::from_edges(&[
+///     (0, 1), (1, 2)
+/// ]);
+/// // Calculate the eigenvector centrality
+/// let output: Result<Option<Vec<f64>>> = katz_centrality(&g, |_| {Ok(1.)}, None, None, None, None, None);
+/// ```
+pub fn katz_centrality<G, F, E>(
+    graph: G,
+    mut weight_fn: F,
+    alpha: Option<f64>,
+    beta_map: Option<HashMap<usize, f64>>,
+    beta_scalar: Option<f64>,
+    max_iter: Option<usize>,
+    tol: Option<f64>,
+) -> Result<Option<Vec<f64>>, E>
+where
+    G: NodeIndexable + IntoNodeIdentifiers + IntoNeighbors + IntoEdges + NodeCount,
+    G::NodeId: Eq + std::hash::Hash,
+    F: FnMut(G::EdgeRef) -> Result<f64, E>,
+{
+    let alpha: f64 = alpha.unwrap_or(0.1);
+
+    let mut beta: HashMap<usize, f64> = beta_map.unwrap_or_else(HashMap::new);
+
+    if beta.is_empty() {
+        // beta_map was none
+        // populate hashmap with default value
+        let beta_scalar = beta_scalar.unwrap_or(1.0);
+        for node_index in graph.node_identifiers() {
+            let node = graph.to_index(node_index);
+            beta.insert(node, beta_scalar);
+        }
+    }
+
+    // Check if beta contains all node indices
+    for node_index in graph.node_identifiers() {
+        let node = graph.to_index(node_index);
+        if !beta.contains_key(&node) {
+            return Ok(None); // beta_map was provided but did not include all nodes
+        }
+    }
```

Comment:
I think putting this in an if/else block would be good here because we don't need to validate `beta` has all the node indices if it was `None`.
```suggestion
    if beta.is_empty() {
        // beta_map was none
        // populate hashmap with default value
        let beta_scalar = beta_scalar.unwrap_or(1.0);
        for node_index in graph.node_identifiers() {
            let node = graph.to_index(node_index);
            beta.insert(node, beta_scalar);
        }
    } else {
        // Check if beta contains all node indices
        for node_index in graph.node_identifiers() {
            let node = graph.to_index(node_index);
            if !beta.contains_key(&node) {
                return Ok(None); // beta_map was provided but did not include all nodes
            }
        }
    }
```

Reply:
Done